File: 2022/day9/day9.py
#!/usr/bin/env python3

from pathlib import Path
import sys

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text()


def show(rope):

    minx = min(k[0] for k in rope)
    maxx = max(k[0] for k in rope)
    miny = min(k[1] for k in rope)
    maxy = max(k[1] for k in rope)

    minx = min(minx, 0)
    miny = min(miny, 0)
    maxx = max(maxx, 5)
    maxy = max(maxy, 4)

    for y in range(maxy, miny - 1, -1):
        row = ""
        for x in range(minx, maxx + 1):
            for i, knot in enumerate(rope):
                if knot == (x, y):
                    label = "H" if i == 0 else "T" if i == len(rope) - 1 else str(i)
                    row += label
                    break
            else:
                if (x, y) == (0, 0):
                    row += "s"
                else:
                    row += "."
        print(row)
    print()


def move(dir, head, tail):
    hx, hy = head
    tx, ty = tail

    # head movement
    if dir == "R":
        dx, dy = 1, 0
    elif dir == "L":
        dx, dy = -1, 0
    elif dir == "U":
        dx, dy = 0, 1
    elif dir == "D":
        dx, dy = 0, -1
    else:
        dx, dy = 0, 0

    hx += dx
    hy += dy

    # tail movement
    dx = 1 if hx - tx > 0 else -1
    dy = 1 if hy - ty > 0 else -1

    if tx == hx:
        if abs(ty - hy) > 1:
            ty += dy
    elif ty == hy:
        if abs(tx - hx) > 1:
            tx += dx
    elif abs(tx - hx) + abs(ty - hy) > 2:
        tx += dx
        ty += dy

    head = (hx, hy)
    tail = (tx, ty)

    return head, tail


def move_rope(dir, rope):
    for k in range(len(rope) - 1):
        rope[k], rope[k + 1] = move(dir if k == 0 else "", rope[k], rope[k + 1])


# part 1
H = (0, 0)
T = (0, 0)
tails = set()
tails.add(T)
for line in data.splitlines():
    dir, n = line.split()
    for i in range(int(n)):
        H, T = move(dir, H, T)
        tails.add(T)
print(len(tails))
# show([H, T])


# part 2
rope = [(0, 0)] * 10
tails = set()
tails.add(rope[-1])
for line in data.splitlines():
    dir, n = line.split()
    for _ in range(int(n)):
        move_rope(dir, rope)
        tails.add(rope[-1])
print(len(tails))
# show(rope)

File: 2022/day9/day9.rs
//! [Day 9: Rope Bridge](https://adventofcode.com/2022/day/9)

use rustc_hash::FxHashSet;

struct Puzzle {
    moves: Vec<(char, i32)>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        Self {
            moves: data
                .lines()
                .map(|line| {
                    let mut split = line.split(' ');
                    let direction = split.next().unwrap().chars().next().unwrap();
                    let n = split.next().unwrap().parse().unwrap();

                    (direction, n)
                })
                .collect(),
        }
    }

    // Solves part one
    fn part1(&self) -> usize {
        let mut tails = FxHashSet::default();
        let mut head = (0, 0);
        let mut tail = (0, 0);

        for m in &self.moves {
            let direction = m.0;
            let n = m.1;
            for _ in 0..n {
                Self::move_segment(direction, &mut head, &mut tail);

                tails.insert(tail);
            }
        }
        tails.len()
    }

    // Solve part two
    fn part2(&self) -> usize {
        let mut tails = FxHashSet::default();
        let mut rope = [
            (0, 0),
            (0, 0),
            (0, 0),
            (0, 0),
            (0, 0),
            (0, 0),
            (0, 0),
            (0, 0),
            (0, 0),
            (0, 0),
        ];

        for m in &self.moves {
            let direction = m.0;
            let n = m.1;
            for _ in 0..n {
                Self::move_rope(direction, &mut rope);
                tails.insert(rope.last().copied().unwrap());
            }
        }
        tails.len()
    }

    /// Move a multi-segment rope
    fn move_rope(direction: char, rope: &mut [(i32, i32)]) {
        for k in 0..rope.len() - 1 {
            let mut head = rope[k];
            let mut tail = rope[k + 1];

            Self::move_segment(if k == 0 { direction } else { ' ' }, &mut head, &mut tail);

            rope[k] = head;
            rope[k + 1] = tail;
        }
    }

    /// Move a segment
    fn move_segment(direction: char, head: &mut (i32, i32), tail: &mut (i32, i32)) {
        // head movement
        let (dx, dy) = match direction {
            'R' => (1, 0),
            'L' => (-1, 0),
            'U' => (0, 1),
            'D' => (0, -1),
            _ => (0, 0),
        };
        head.0 += dx;
        head.1 += dy;

        // tail movement
        let dx = if head.0 - tail.0 > 0 { 1 } else { -1 };
        let dy = if head.1 - tail.1 > 0 { 1 } else { -1 };
        if tail.0 == head.0 {
            if (tail.1 - head.1).abs() > 1 {
                tail.1 += dy;
            }
        } else if tail.1 == head.1 {
            if (tail.0 - head.0).abs() > 1 {
                tail.0 += dx;
            }
        } else if (tail.0 - head.0).abs() + (tail.1 - head.1).abs() > 2 {
            tail.0 += dx;
            tail.1 += dy;
        }
    }
}

#[must_use]
pub fn solve(data: &str) -> (usize, usize) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");
    const TEST_INPUT_2: &str = include_str!("test02.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 13);
        assert_eq!(puzzle.part2(), 1);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(TEST_INPUT_2);
        assert_eq!(puzzle.part2(), 36);
    }
}
